libs/partners/xai/langchain_xai/chat_models.py

```diff
@@ -0,0 +1,351 @@
+                },
+                {
+                    'name': 'GetPopulation',
+                    'args': {'location': 'LA'},
+                    'id': 'call_0vjgq455gq1av5sp9eb1pw6a',
+                    'type': 'tool_call'
+                }
+            ]
+
+    Structured output:
+        .. code-block:: python
+
+            from typing import Optional
+
+            from pydantic import BaseModel, Field
+
+
+            class Joke(BaseModel):
+                '''Joke to tell user.'''
+
+                setup: str = Field(description="The setup of the joke")
+                punchline: str = Field(description="The punchline to the joke")
+                rating: Optional[int] = Field(description="How funny the joke is, from 1 to 10")
+
+
+            structured_llm = llm.with_structured_output(Joke)
+            structured_llm.invoke("Tell me a joke about cats")
+
+        .. code-block:: python
+
+            Joke(
+                setup='Why was the cat sitting on the computer?',
+                punchline='To keep an eye on the mouse!',
+                rating=7
+            )
+
+    Token usage:
+        .. code-block:: python
+
+            ai_msg = llm.invoke(messages)
+            ai_msg.usage_metadata
+
+        .. code-block:: python
+
+            {'input_tokens': 37, 'output_tokens': 6, 'total_tokens': 43}
+
+    Logprobs:
+        .. code-block:: python
+
+            logprobs_llm = llm.bind(logprobs=True)
+            messages=[("human","Say Hello World! Do not return anything else.")]
+            ai_msg = logprobs_llm.invoke(messages)
+            ai_msg.response_metadata["logprobs"]
+
+        .. code-block:: python
+
+            {
+                'content': None,
+                'token_ids': [22557, 3304, 28808, 2],
+                'tokens': [' Hello', ' World', '!', '</s>'],
+                'token_logprobs': [-4.7683716e-06, -5.9604645e-07, 0, -0.057373047]
+            }
+
+
+    Response metadata
+        .. code-block:: python
+
+            ai_msg = llm.invoke(messages)
+            ai_msg.response_metadata
+
+        .. code-block:: python
+
+            {
+                'token_usage': {
+                    'completion_tokens': 4,
+                    'prompt_tokens': 19,
+                    'total_tokens': 23
+                    },
+                'model_name': 'grok-beta',
+                'system_fingerprint': None,
+                'finish_reason': 'stop',
+                'logprobs': None
+            }
+
+    """  # noqa: E501
+
+    model_name: str = Field(alias="model")
+    """Model name to use."""
+    xai_api_key: Optional[SecretStr] = Field(
+        alias="api_key",
+        default_factory=secret_from_env("XAI_API_KEY", default=None),
+    )
+    """xAI API key.
+
+    Automatically read from env variable `XAI_API_KEY` if not provided.
+    """
+    xai_api_base: str = Field(
+        default_factory=from_env("XAI_API_BASE", default="https://api.x.ai/v1/"),
```

Comment:
nit: imo doesn't need to be env configurable

Reply:
can always introduce later if it'd be useful, but would prefer to minimize num of env vars we have so there's fewer backwards compatibility considerations in future